clients/drcachesim/drpt2trace/pt2ir.h

```diff
@@ -0,0 +1,272 @@
+};
+
+struct pt_image_section_cache;
+struct pt_sb_pevent_config;
+struct pt_sb_session;
+struct pt_insn_decoder;
+struct pt_ins;
+
+/**
+ * pt2ir_t is a class that can convert PT raw trace to DynamoRIO's IR.
+ */
+class pt2ir_t {
+public:
+    pt2ir_t();
+    ~pt2ir_t();
+
+    /**
+     * Returns TRUE if the pt2ir_t is successfully initialized. Returns FALSE on failure.
+     * \note Parse struct pt2ir_config_t and initialize PT instruction decoder, the
+     * sideband session, and images caches.
+     */
+    bool
+    init(IN const pt2ir_config_t pt2ir_config);
+
+    /**
+     * Returns pt2ir_convert_status_t.
+     * \note The convert function performs two processes: (1) decode the PT raw trace into
+     * libipt's IR format pt_insn; (2) convert pt_insn into the doctor's IR format
+     * instr_t. If the convertion is successful, the function returns PT2IR_CONV_SUCCESS.
+     * Otherwise, the function returns the corresponding error code.
+     */
+    pt2ir_convert_status_t
+    convert();
+
+    /**
+     * Returns all converted DynamoRIO IR.
+     */
+    std::vector<instr_t>
+    get_instrlist();
+
+    /**
+     * Returns the number of instructions in the converted IR.
+     */
+    uint64_t
+    get_instr_count();
+
+    /**
+     * Print the disassembled text of all valid instructions to STDOUT.
+     */
+    void
+    print_instrs_to_stdout();
+
+private:
+    /* Load PT raw file to buffer. The struct pt_insn_decoder will decode this buffer to
+     * libipt's IR.
+     */
+    bool
+    load_pt_raw_file(IN std::string path);
+
+    /* Load the elf section in kcore to sideband session iscache and store the section
+     * index to sideband kimage.
+     * XXX: Could we not use kcore? We can store all kernel modules and the mapping
+     * information.
+     */
+    bool
+    load_kernel_image(IN std::string path);
+
+    /* Allocate a sideband decoder in the sideband session. The sideband session may
+     * allocate many decoders, which mainly work on handling sideband perf records and
+     * help the PT decoder switch images. */
+    bool
+    alloc_sb_pevent_decoder(IN struct pt_sb_pevent_config *config);
+
+    /* Diagnose converting errors and output diagnostic results.
+     * It will used to generate the error message during the decoding process.
+     */
+    void
+    dx_decoding_error(IN int errcode, IN const char *errtype, IN uint64_t ip);
+
+    /* Buffer for caching the PT raw trace. */
+    void *pt_raw_buffer_;
+    size_t pt_raw_buffer_size_;
+
+    /* The libipt instruction decoder. */
+    struct pt_insn_decoder *pt_instr_decoder_;
+
+    /* The libipt image section cache. */
+    struct pt_image_section_cache *pt_iscache_;
+
+    /* The libipt sideband session. */
+    struct pt_sb_session *pt_sb_session_;
+
+    /* All valid instructions that can be decoded from PT raw trace. */
+    std::vector<struct pt_insn> pt_instr_list_;
+
+    /* The DynamoRIO's IR list. */
+    std::vector<instr_t> instr_list_;
```

Comment:
Unless we need `instr_list_` somewhere else too, we can probably have this as a local in get_instrlist

Reply:
Removed. I think we can discuss it on next PR.